### ISSUE TITLE
apt-get updates appear to be running slower now

### ISSUE THREAD

**Issue by tfoote** [ros-infrastructure/ros_buildfarm#535]
We've observed noteable degraded performance on apt-get updates on the farm. It's causing some jobs to run for more than 10x their original runtimes. 

http://build.ros.org/view/Ibin_uT32/job/Ibin_uT32__jsk_2016_01_baxter_apc__ubuntu_trusty_i386__binary/
http://build.ros.org/view/Ibin_uT32/job/Ibin_uT32__pr2_gripper_sensor_controller__ubuntu_trusty_i386__binary/
One of them failed 3 days in a row then worked. The other failed multiple times in a row and they seem to be timing out in the apt-get install stage.

The pr2 one used to be 10 - 15 minute jobs but the one recovered took a full hour. http://build.ros.org/view/Ibin_uT32/job/Ibin_uT32__pr2_gripper_sensor_controller__ubuntu_trusty_i386__binary/30/

From @nuclearsandwich 
I've noticed anecdotally (especially on my testfarms) that apt seems to be taking longer lately. I don't know of anything offhand but checking mirror status and the health/utilization of the squid proxies can go on my list for next week.

And @mikaelarguedas has also noticed issues with armhf builds that already go slower so the changes are more noticeable.

Here's an example of a ridiculously long apt operation, taking close to an hour from: http://build.ros.org/view/Ibin_uT32/job/Ibin_uT32__pr2_gripper_sensor_controller__ubuntu_trusty_i386__binary/29/consoleFull
```
00:06:03.399 Get:190 http://archive.ubuntu.com/ubuntu/ trusty-updates/main unzip i386 6.0-9ubuntu1.5 [150 kB]
00:06:03.701 Get:191 http://archive.ubuntu.com/ubuntu/ trusty/main wdiff i386 1.2.1-2 [28.0 kB]
00:06:03.995 Get:192 http://archive.ubuntu.com/ubuntu/ trusty/main libauthen-sasl-perl all 2.1500-1 [53.2 kB]
00:06:04.288 Get:193 http://archive.ubuntu.com/ubuntu/ trusty-updates/main python3-magic all 1:5.14-2ubuntu3.3 [4,570 B]
00:14:57.456 W: Duplicate sources.list entry http://archive.ubuntu.com/ubuntu/ trusty/multiverse i386 Packages (/var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty_multiverse_binary-i386_Packages.gz)
00:14:57.456 W: Duplicate sources.list entry http://archive.ubuntu.com/ubuntu/ trusty-updates/multiverse i386 Packages (/var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty-updates_multiverse_binary-i386_Packages.gz)
00:23:16.715 
Extracting templates from packages: 15%W: Duplicate sources.list entry http://archive.ubuntu.com/ubuntu/ trusty/multiverse i386 Packages (/var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty_multiverse_binary-i386_Packages.gz)
00:23:16.715 W: Duplicate sources.list entry http://archive.ubuntu.com/ubuntu/ trusty-updates/multiverse i386 Packages (/var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty-updates_multiverse_binary-i386_Packages.gz)
00:31:14.616 
Extracting templates from packages: 31%W: Duplicate sources.list entry http://archive.ubuntu.com/ubuntu/ trusty/multiverse i386 Packages (/var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty_multiverse_binary-i386_Packages.gz)
00:31:14.616 W: Duplicate sources.list entry http://archive.ubuntu.com/ubuntu/ trusty-updates/multiverse i386 Packages (/var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty-updates_multiverse_binary-i386_Packages.gz)
00:38:53.723 
Extracting templates from packages: 46%W: Duplicate sources.list entry http://archive.ubuntu.com/ubuntu/ trusty/multiverse i386 Packages (/var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty_multiverse_binary-i386_Packages.gz)
00:38:53.723 W: Duplicate sources.list entry http://archive.ubuntu.com/ubuntu/ trusty-updates/multiverse i386 Packages (/var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty-updates_multiverse_binary-i386_Packages.gz)
00:46:34.496 
Extracting templates from packages: 62%W: Duplicate sources.list entry http://archive.ubuntu.com/ubuntu/ trusty/multiverse i386 Packages (/var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty_multiverse_binary-i386_Packages.gz)
00:46:34.496 W: Duplicate sources.list entry http://archive.ubuntu.com/ubuntu/ trusty-updates/multiverse i386 Packages (/var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty-updates_multiverse_binary-i386_Packages.gz)
00:54:13.369 
Extracting templates from packages: 77%W: Duplicate sources.list entry http://archive.ubuntu.com/ubuntu/ trusty/multiverse i386 Packages (/var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty_multiverse_binary-i386_Packages.gz)
00:54:13.369 W: Duplicate sources.list entry http://archive.ubuntu.com/ubuntu/ trusty-updates/multiverse i386 Packages (/var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty-updates_multiverse_binary-i386_Packages.gz)
00:58:00.284 
Extracting templates from packages: 93%W: Duplicate sources.list entry http://archive.ubuntu.com/ubuntu/ trusty/multiverse i386 Packages (/var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty_multiverse_binary-i386_Packages.gz)
00:58:00.285 W: Duplicate sources.list entry http://archive.ubuntu.com/ubuntu/ trusty-updates/multiverse i386 Packages (/var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_trusty-updates_multiverse_binary-i386_Packages.gz)
00:58:00.286 
Extracting templates from packages: 100%
00:58:00.296 Preconfiguring packages ...
00:58:00.537 Fetched 65.5 MB in 47s (1,375 kB/s)
```

There are a possibly related articles we found that might suggest it's an issue with new ipv6 rollouts:
https://www.techrepublic.com/article/how-to-fix-the-slow-apt-get-update-issue-on-linux-machines/

And we might be able to lower some timeouts: https://askubuntu.com/questions/141513/how-to-lower-wait-time-for-repository-updates

I can reproduce slow updates on a local docker instance too, though not as bad:
```
$ docker run -ti --rm ubuntu:trusty bash
root@e775cc81ceab:/# time apt-get update
Get:1 http://security.ubuntu.com trusty-security InRelease [65.9 kB]       
Ign http://archive.ubuntu.com trusty InRelease                                 
Get:2 http://archive.ubuntu.com trusty-updates InRelease [65.9 kB]            
Get:3 http://security.ubuntu.com trusty-security/universe Sources [85.4 kB]    
Get:4 http://archive.ubuntu.com trusty-backports InRelease [65.9 kB]           
Get:5 http://archive.ubuntu.com trusty Release.gpg [933 B]                     
Get:6 http://security.ubuntu.com trusty-security/main amd64 Packages [901 kB]  
Get:7 http://archive.ubuntu.com trusty Release [58.5 kB]                       
Get:8 http://security.ubuntu.com trusty-security/restricted amd64 Packages [18.1 kB]
Get:9 http://security.ubuntu.com trusty-security/universe amd64 Packages [283 kB]
Get:10 http://security.ubuntu.com trusty-security/multiverse amd64 Packages [4717 B]
Get:11 http://archive.ubuntu.com trusty-updates/universe Sources [250 kB]      
Get:12 http://archive.ubuntu.com trusty-updates/main amd64 Packages [1324 kB]  
Get:13 http://archive.ubuntu.com trusty-updates/restricted amd64 Packages [21.4 kB]
Get:14 http://archive.ubuntu.com trusty-updates/universe amd64 Packages [579 kB]
Get:15 http://archive.ubuntu.com trusty-updates/multiverse amd64 Packages [16.0 kB]
Get:16 http://archive.ubuntu.com trusty-backports/main amd64 Packages [14.7 kB]
Get:17 http://archive.ubuntu.com trusty-backports/restricted amd64 Packages [40 B]
Get:18 http://archive.ubuntu.com trusty-backports/universe amd64 Packages [52.5 kB]
Get:19 http://archive.ubuntu.com trusty-backports/multiverse amd64 Packages [1392 B]
Get:20 http://archive.ubuntu.com trusty/universe Sources [7926 kB]             
Get:21 http://archive.ubuntu.com trusty/main amd64 Packages [1743 kB]          
Get:22 http://archive.ubuntu.com trusty/restricted amd64 Packages [16.0 kB]    
Get:23 http://archive.ubuntu.com trusty/universe amd64 Packages [7589 kB]      
Get:24 http://archive.ubuntu.com trusty/multiverse amd64 Packages [169 kB]     
Fetched 21.3 MB in 2min 26s (145 kB/s)                                         
Reading package lists... Done

real	2m47.258s
user	0m5.812s
sys	0m36.348s
```

**Comment by gavanderhoorn** [comment]
It's anecdotal, and may be a red herring, but I've run into this the past few weeks as well, and after disabling spectre and meltdown patches/work-around (ie: `nopti`, `nospectre_v2`) I've noticed a significant performance increase. Especially for IO bound tasks.

I've not done any proper verification of whether these two kernel features really contribute, but seeing as most Docker image builds have IO bound phases in them, I just wanted to mention it.

**Comment by nuclearsandwich** [comment]
Looking at this today running docker containers on machines in AWS comparable to the build.ros.org agent nodes.

`time apt-get update` inside a fresh Docker container seems to run about an order of magnitude slower for trusty than for xenial. ~40s realtime for trusty versus ~4s for xenial. Nearly all of the scheduled time was system time rather than user time. Which is why the Fetched... summary in the apt-get output disagrees so much with the actual runtime, it seems to account only for user time or some subset of it.

To run strace on apt-get and see where the time was going (the vast majority was system time on trusty) I started running docker containers with `--privileged`. Surprisingly, this reduced the realtime to ~10s. So still slower than xenial but not by nearly as much. Adding only the SYS_PTRACE capability to the container is enough to enable strace usage without affecting the performance we're trying to trace but based on that accidental discovery it does seem like there's some system capability that apt-get for trusty is expecting to be there.

strace showed a handful of long-blocking read and select calls but I haven't analyzed the full trace yet.

**Comment by nuclearsandwich** [comment]
Tried to reproduce this on a local environment today and I wasn't able to exhibit the issue on my workstation which runs docker 18.02. The docker puppet module hadn't made the jump to the new docker apt repo and so it is still running 17.05 I'm currently spinning up a new test environment to see if I can reproduce the issue on an agent-sized host in AWS with the latest (18.03) version of docker. If not, we'll figure out the upgrade path and move the discussion to the buildfarm_deployment repository.

**Comment by nuclearsandwich** [comment]
If anyone else wants to take a look at the strace outputs here are a few samples taken with `strace -T`. https://gist.github.com/nuclearsandwich/7c36a9db427cc98dc99f8d1db56bc19f

**Comment by tfoote** [comment]
Here's a blog post talking about how to profile performance issues in docker. https://hackernoon.com/another-reason-why-your-docker-containers-may-be-slow-d37207dec27f I doubt we have exactly the same problem, but the debugging approach might be effective.

**Comment by nuclearsandwich** [comment]
Trying to get a better handle on this, I pulled up a bunch of the trend graphs from Indigo to get a better handle on when this started. Bad screenshots below for posterity.

![ibin_ut64_jsk_pcl_ros_trend](https://user-images.githubusercontent.com/358882/38957948-db302cb2-4329-11e8-815a-cc4fc7c61cc0.png)
![ibin_ut64_jsk_perception_trend](https://user-images.githubusercontent.com/358882/38957949-db411edc-4329-11e8-8adf-771a571ba701.png)
![ibin_ut64__master_sync_fkie_trend](https://user-images.githubusercontent.com/358882/38957950-db53fff2-4329-11e8-9841-50f7ffd031c2.png)

They don't tell the story on their own but it looks the last builds before times spiked were between March 19 and March 21. March 21 is when I rebuilt the agent image to deploy the new Docker seccomp policy for https://github.com/ros-infrastructure/buildfarm_deployment/issues/176 so that image build is a strong contender for precipitating event.

I have to collect more data but having reverted to the default seccomp profile on a few hosts and still seeing extreme times for apt-get, I don't believe that it's the (only?) issue.

The previous images had `uname -a`
`Linux ... 4.4.0-112-generic #135-Ubuntu SMP Fri Jan 19 11:48:36 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`
and the current ones feature
`Linux ... 4.4.0-1052-aws #61-Ubuntu SMP Mon Feb 12 23:05:58 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`
I'm also reproducing the issue locally with an HWE kernel from Jan 25.
`Linux steven-workstation 4.13.0-32-generic #35~16.04.1-Ubuntu SMP Thu Jan 25 10:13:43 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`

If the issue is indeed related to spectre/meltdown mitigations we'll have to figure out what path forward to take.

**Comment by nuclearsandwich** [comment]
I've confirmed that the seccomp policy change alone has no significant effect on apt-get timing in trusty containers.

**Comment by tfoote** [comment]
There are quite a few builds that have jumped up to just below our threshold:

http://build.ros.org/view/Ibin_uT32/job/Ibin_uT32__constrained_ik__ubuntu_trusty_i386__binary/buildTimeTrend
![image](https://user-images.githubusercontent.com/447804/39149247-58333d90-46f3-11e8-858d-8a1272903faf.png)

http://build.ros.org/view/Ibin_uT32/job/Ibin_uT32__nextage_gazebo__ubuntu_trusty_i386__binary/buildTimeTrend
![image](https://user-images.githubusercontent.com/447804/39149287-754e9b72-46f3-11e8-92cd-3d11454636ac.png)

both of which jumped to about 4x longer on Trusty 32

**Comment by nuclearsandwich** [comment]
I was able to confirm what @gavanderhoorn mentioned about the spectre and meltdown mitigations. Disabling them does resolve the apt-get performance issues in Trusty containers running on Xenial hosts with the linux-aws kernel.

Since I would like to avoid disabling those mitigations on build.ros.org I looked at other options: running Trusty containers on Trusty hosts does not exhibit the issue and as of a test just now, neither did running Trusty containers on Bionic hosts. Kernel uname -a

```
Linux HOSTNAME 4.15.0-1007-aws #7-Ubuntu SMP Tue Apr 24 10:56:17 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

So if we cannot identify the issue clearly enough to resolve it for Xenial hosts we have the option to move forward as well as backward on the host systems.

**Comment by tfoote** [comment]
Also of note. This has been identified as effecting amd64 Trusty builds as well. It seems to cause a 3x slowdown not the 4x slowdown on i386. And there's also been devel jobs that have timed out recently that were not near an issue before either.

**Comment by nuclearsandwich** [comment]
The issue is also resolved in the 4.15.0-20-generic kernel available in xenial on AWS in the package `linux-virtual-hwe-16.04-edge`. I don't know at the moment the full list of optimizations we sacrifice to switch from the AWS-tuned 4.4 kernel to the generic 4.15 but at the very least the ENA network driver is included and still appears to function on this kernel.

I think switching our production kernel for agent hosts is the way we'll fix this issue on build.ros.org, but we don't make kernel modifications in puppet so there's nothing that we would necessarily change in the source here except to add documentation for this known limitation.

**Comment by nuclearsandwich** [comment]
[pr2_gripper_sensor_controller](http://build.ros.org/view/Ibin_uT32/job/Ibin_uT32__pr2_gripper_sensor_controller__ubuntu_trusty_i386__binary/buildTimeTrend) which was timing out after two hours now has 1hr:20min builds when the docker image needs to be refreshed and 11min builds with a ready image.

#35 | 11 min | agent-34aa77e6
#34 | 11 min | agent-e856716b
#33 | 1 hr 20 min | agent-921c15d4

Neither of the other jobs mentioned [here](https://github.com/ros-infrastructure/ros_buildfarm/issues/535#issuecomment-383697766) have been successfully rebuilt. Which could be a result of missing dependencies due to https://github.com/ros-infrastructure/buildfarm_deployment/issues/204

**Comment by nuclearsandwich** [comment]
As of yesterday we've rolled back the fix for this performance issue as it caused worse issues with x86-32 Indigo builds.

**Comment by gavanderhoorn** [comment]
Might be something to check: https://github.com/docker/for-linux/issues/247#issuecomment-404116355.

**Comment by gavanderhoorn** [comment]
According to https://github.com/docker/for-linux/issues/247#issuecomment-434204454, setting the IO scheduler to `noop` worked to increase performance (but in 18.04 containers). Might be something to check out if we're still having issues.

**Comment by nuclearsandwich** [comment]
> What is the status on this?

Nothing has changed since the [update on June 5](https://github.com/ros-infrastructure/ros_buildfarm/issues/535#issuecomment-394707038). We're living with the performance issues for now.

**Comment by dirk-thomas** [comment]
> Nothing has changed since the update on June 5.

Ok, I only checked since the ticket is assigned to you and I was wondering if you are working or planning to works on this soon...

**Comment by nuclearsandwich** [comment]
It's not anything I have scheduled. Since the issue is with infrastructure rather than anything in the ros_buildfarm tooling I could use GitHub's handy dandy new issue mover to try and move the issue to buildfarm_deployment. Or we could close the issue here with an imperfect resolution.

---

The last time @tfoote and I discussed it we decided that we wanted to preserve the Spectre and Meltdown mitigations rather than try disabling them without taking full time to explore the ramifications.

@tfoote and I also recently made an adjustment to the IO provisioning which might contribute to a partial mitigation. Of the sample jobs that previously exhibited issues I see that [constrained_ik](http://build.ros.org/view/Ibin_uT32/job/Ibin_uT32__constrained_ik__ubuntu_trusty_i386__binary/buildTimeTrend) has recent builds that are only 5-10 minutes longer than prior to the perf hit.
